extensions/infinispan-client/deployment/src/main/java/io/quarkus/infinispan/client/deployment/InfinispanClientProcessor.java

```diff
@@ -281,4 +463,160 @@ void registerServiceBinding(Capabilities capabilities, BuildProducer<ServiceProv
+                            && i.getRequiredQualifier(INFINISPAN_CLIENT_ANNOTATION) == null)
+                    .findAny()
+                    .isPresent();
+            if (createDefaultCacheManager) {
+                clientNames.add(DEFAULT_INFINISPAN_CLIENT_NAME);
+            }
+        }
+
+        // Produce default and/or named RemoteCacheManager and CounterManager beans
+        for (String clientName : clientNames) {
+            syntheticBeanBuildItemBuildProducer.produce(
+                    configureAndCreateSyntheticBean(clientName, RemoteCacheManager.class,
+                            recorder.infinispanClientSupplier(clientName)));
+            syntheticBeanBuildItemBuildProducer.produce(
+                    configureAndCreateSyntheticBean(clientName, CounterManager.class,
+                            recorder.infinispanConterManagerSupplier(clientName)));
+        }
+
+        // Produce RemoteCache beans
+        for (RemoteCacheBean remoteCacheBean : remoteCacheBeans) {
+            syntheticBeanBuildItemBuildProducer.produce(
+                    configureAndCreateSyntheticBean(remoteCacheBean,
+                            recorder.infinispanRemoteCacheClientSupplier(remoteCacheBean.clientName,
+                                    remoteCacheBean.cacheName)));
+        }
+    }
+
+    static <T> SyntheticBeanBuildItem configureAndCreateSyntheticBean(String name,
+            Class<T> type,
+            Supplier<T> supplier) {
+        SyntheticBeanBuildItem.ExtendedBeanConfigurator configurator = buildSyntheticBean(
+                SyntheticBeanBuildItem.configure(type).supplier(supplier));
+
+        if (InfinispanClientUtil.isDefault(name)) {
+            configurator.addQualifier(Default.class);
+        } else {
+            configurator.addQualifier().annotation(DotNames.NAMED).addValue("value", name).done();
+            configurator.addQualifier().annotation(INFINISPAN_CLIENT_ANNOTATION).addValue("value", name).done();
+        }
+
+        return configurator.done();
+    }
+
+    static <T> SyntheticBeanBuildItem configureAndCreateSyntheticBean(RemoteCacheBean remoteCacheBean, Supplier<T> supplier) {
+        SyntheticBeanBuildItem.ExtendedBeanConfigurator configurator = SyntheticBeanBuildItem.configure(RemoteCache.class)
+                .types(remoteCacheBean.type)
+                .scope(Singleton.class) // Some Infinispan API won't work if this is not a Mock
+                .supplier(supplier)
+                .unremovable()
+                .setRuntimeInit();
+
+        configurator.addQualifier().annotation(INFINISPAN_REMOTE_ANNOTATION).addValue("value", remoteCacheBean.cacheName)
+                .done();
+
+        configurator.addQualifier().annotation(INFINISPAN_CLIENT_ANNOTATION).addValue("value", remoteCacheBean.clientName)
+                .done();
+        return configurator.done();
+    }
+
+    static <T> SyntheticBeanBuildItem.ExtendedBeanConfigurator buildSyntheticBean(
```

Comment:
This method only seems to be used from within `configureAndCreateSyntheticBean` - I'd probably just inline the code there.

Reply:
oki